Add extra index to package

jialuw96 opened the pull request:
Issue: https://github.com/dowlinglab/sorption-fixed-bed/issues/38

jialuw96 commented:
@adowling2 Please see the newest commit to fim_doe.py of adding extra index updates. Please see the comment above for changes I made to fit extra index in this package. Thank you!

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -122,11 +121,12 @@ def __check_inputs(self, check_mode=False):
+                     jac_involved_measurement=None,
                      scale_nominal_param_value=False, scale_constant_value=1, if_Cholesky=False, L_LB = 1E-10, L_initial=None,
-                     jac_initial=None, fim_initial=None, trace_initial=None, det_initial=None,
+                     jac_initial=None, fim_initial=None,
```
Please note in the future (separate issue?) that L_initial should be calculated from fim_initial. We need to provide an example of using the sequential mode to calculate `jac_initial` and `fim_initial`.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -182,14 +183,21 @@ def optimize_doe(self,  design_values, if_optimize=True, objective_option='det',
+        if jac_involved_measurement is None:
+            m = self.__create_doe_model()
+        else:
+            m = self.__create_doe_model(jac_involved_name = jac_involved_name_, jac_involved_extra_index= jac_involved_extra_index_)
```
Let's talk through how to simplify this today.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -332,7 +344,8 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
                 prior_in_use = specified_prior
 
-            FIM_analysis = FIM_result(self.param_name, self.measurement_variables, self.measurement_timeset, self.measurement_extra_index,  prior_FIM=prior_in_use, store_FIM=FIM_store_name, scale_constant_value=self.scale_constant_value)
+            FIM_analysis = FIM_result(self.param_name, self.measurement_variables, self.measurement_timeset, flatten_all_measure=self.flatten_measure_name,
```
I'd like to see how `flatten_all_measure` is structured in our discussion today.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -735,12 +783,27 @@ def __finite_calculation(self, output_record, scena_gen):
+        '''
+        # 3-D array form of jacobian
+        jac_3Darray = np.zeros((len(self.flatten_measure_name), len(self.param_name), len(self.measurement_timeset[0])))
+        no_time = len(self.measurement_timeset[0])
```
Why is this accessing element 0? This is something to explain in the comments.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -966,12 +1029,14 @@ def run_grid_search(self, design_values, design_ranges, design_dimension_names,
 
         Parameters:
         -----------
+        jac_involved_name: a list of measurement names involved in calculation of FIM
```
Can we simplify this interface?

adowling2 replied:
Let's discuss this on a whiteboard.

adowling2 replied:
Waiting to update this.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -1015,6 +1080,20 @@ def __create_doe_model(self):
         m.t_set = Set(initialize=time_set)
         m.tmea_set = Set(initialize=self.measurement_timeset[0])
 
+        # reorganize measurement names with
```
Should we add a private method named `flatten_measurements`? This would allow us to test it better.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -1035,17 +1114,30 @@ def __create_doe_model(self):
-                for j, un in enumerate(m.para_set):
-                    for t, tim in enumerate(m.tmea_set):
-                        dict_jac[(bu,un,tim)] = self.jac_initial[i,j,t]
+            if jac_involved_name is not None:
```
Simplify this? You could instead say:
```
if jac_involved_name is not None:
   measurements = m.involved_y_set
else:
   measurements = m.y_set
```

Then use `measurements` in the for loops. But I do not fully appreciate why this logic is needed.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -1035,17 +1114,30 @@ def __create_doe_model(self):
-            m.jac = Var(m.y_set, m.para_set, m.tmea_set, initialize=jac_initialize)
+
+            if jac_involved_name is not None:
+                m.jac = Var(m.involved_y_set, m.para_set, m.tmea_set, initialize=jac_initialize)
```
See my comment above. There has to be a better way to do this.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -1196,10 +1314,17 @@ def cholesky_imp(m,c,d):
-        m.dC_value = Constraint(m.y_set, m.para_set, m.tmea_set, rule=jac_numerical)
-        m.ele_rule = Constraint(m.para_set, m.para_set, rule=calc_FIM)  
+        if jac_involved_name is None:
+            m.dC_value = Constraint(m.y_set, m.para_set, m.tmea_set, rule=jac_numerical)
```
See comment above.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -1633,21 +1758,22 @@ def __init__(self, para_name, measurement_variables, measurement_timeset,  prior
         self.verbose = verbose
 
-    def calculate_FIM(self, jaco_information, dv_values, jaco_involved_name=None, jaco_involved_extra_index=None, result=None):
+    def calculate_FIM(self, jaco_information, jaco_3D, dv_values, jaco_involved_name=None, jaco_involved_extra_index=None, result=None):
```
Is jaco_info and jaco_3D redundant? Do you need both?

adowling2 replied:
We should standardize on one general representation for the Jacobian that works for all use cases. This will make the code much easier to maintain.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -1672,25 +1798,29 @@ def calculate_FIM(self, jaco_information, dv_values, jaco_involved_name=None, ja
-                                if ind in jaco_involved_extra_index:
-                                    jaco_parameter.append(jaco_information[par])[flatten_measure_count]
+            involved_flatten_index = []
+            for n, nam in enumerate(jaco_involved_name):
```
We should centralize all code related to measurement flattening.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -1775,7 +1906,30 @@ def extract_FIM(self, m, dv_set, result, y_set=None, t_all_set=None, obj=None, a
                         JAC[n1, n2, t] = value(m.jac[name1, name2, tim])
         print('JAC is:', JAC)
+
+        jaco_info = {}
```
You see to be duplicating code. This makes it harder to maintain code in the long term. Instead, we should generalize this and move it to a function.

adowling2 replied:
Let's talk about how to generalize the remove the duplicated code.

adowling2 reviewed (commented):
Topics for meeting today:
- Agree on the interface to specific measurements
- Generalize code for flattening measurements, converting between Jacobian representations
- Discuss location of code

jialuw96 commented:
@adowling2  Please see updates on fim_doe.py. Main changes include a measurement information class is added to store and pass measurement information, and update optimize_doe() function and create_doe_model() function to make flattened measurement name the general and default choice. Thank you!

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,11 +6,84 @@
+from pyomo.contrib.sensitivity_toolbox.sens import sipopt, sensitivity_calculation, get_dsdp
 
-class DesignOfExperiments: 
+class Measurement_flatten:
```
Change the name to `Measurements`

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,11 +6,84 @@
+        Parameters:
+        ----------
+        measurement_timeset: a dictionary, keys are measurement names in the model, values are a list of time they are measured
+        measurement_extra_index: A dictionary, keys are the measurement names, values are a list of indexes. If no extra indexes, values are None.
```
Define a motivating example here in the comments. This should be an example we contribute to the Pyomo repo.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,11 +6,84 @@
+    def flatten_measurement(self):
+        '''Flatten measurements and their extra index (if any)
+        Flattened names consist of their original names in the model and their extra index value, connected with a '_index_'
+        For e.g., FCO2 with an extra index 19 is 'FCO2_index_19', FCO2 without extra index is still 'FCO2'
```
Replace this with the motivating example.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,11 +6,84 @@
+        for j in list(jac_involved.keys()):
+            if jac_involved[j] is not None: # if it has extra index
+                for ind in jac_involved[j]:
+                    flatten_name = j + '_index_' + str(ind)
```
It looks like this is repeated code. How can we generalize it?

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,11 +6,84 @@
+
+        Parameters
+        ----------
+        jac_involved: a dictionary, keys are measurement names, values are lists of extra indexes
```
`jac_involved` is not a good name. We need a more descriptive name here.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,11 +6,84 @@
+                for ind in jac_involved[j]:
+                    flatten_name = j + '_index_' + str(ind)
+                    jac_involved_name.append(flatten_name)
+            else:
```
Add comment here to describe what `else` means

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,11 +6,84 @@
+        self.flatten_measure_timeset = flatten_measure_timeset
+        return flatten_measure_name, flatten_measure_timeset
+
+    def partly_flatten_measurement(self, jac_involved):
```
Is this method needed anymore? Also, should this be a static method? Does it use any information in self?

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,11 +6,84 @@
+        return jac_involved_name
+
+class DesignOfExperiments:
     def __init__(self, param_init, design_variable_timepoints, measurement_variable_timepoints, create_model, solver=None,
```
This should take a measurement object.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,11 +6,84 @@
+        self.measurement_extra_index = measurement_extra_index
+        # which measurement has extra indexes
+        self.extra_measure_name = list(measurement_extra_index.keys())
+
```
We should add logic here to test if the measurement object is valid. We should also add logic to test if a measurement object is a subset of another measurement object.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -49,32 +120,15 @@ def __init__(self, param_init, design_variable_timepoints, measurement_variable_
 
-        self.flatten_measure_name = flatten_measure_name
+        # create the measurement information object
+        self.measure = Measurement_flatten(measurement_variable_timepoints, measurement_extra_index)
```
Alternate design: the user provides a valid measurement object. Information such as `flatten_measurement_name` and `flatten_measurement_timeset` should be stored in the measurement object.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -122,8 +176,9 @@ def __check_inputs(self, check_mode=False):
 
 
     def optimize_doe(self,  design_values, if_optimize=True, objective_option='det',
```
Alternate design: this optionally takes a measurement object instead of `jac_involvement_measurement`. Or this functionality moves to the measurement class. I think we should isolate all of the logic associated with flattening inside the measurement class.

adowling2 replied:
Still need to update `optimize_doe`.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -137,11 +192,14 @@ def optimize_doe(self,  design_values, if_optimize=True, objective_option='det',
         design_values: initial point for optimization, a dict whose keys are design variable names, values are a dict whose keys are time point and values are the design variable value at that time point
         if_optimize: if true, continue to do optimization. else, just run square problem with given design variable values
         objective_option: supporting maximizing the 'det' determinant or the 'trace' trace of the FIM
+        jac_involved_measurement: a dictionary of measurement names and extra indexes. It should be subsets of the measurement sets given to this class.
```
Let's discuss how to move this into the measurement object.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -137,11 +192,14 @@ def optimize_doe(self,  design_values, if_optimize=True, objective_option='det',
         if_Cholesky: if true, cholesky decomposition is used for Objective function (to optimize determinant).
             L_LB: if FIM is P.D., the diagonal element should be positive, so we can set a LB like 1E-10
             L_initial: initialize the L
+        jac_initial: a matrix used to initialize jacobian matrix
```
In the doc string, explain the default behavior.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -182,6 +238,13 @@ def optimize_doe(self,  design_values, if_optimize=True, objective_option='det',
         self.fim_scale_constant_value = self.scale_constant_value **2
 
+        # identify measurements involved in calculation
+        if jac_involved_measurement is not None:
```
Let's talk about how to streamline this.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -729,18 +821,17 @@ def __finite_calculation(self, output_record, scena_gen):
             list_jac = []
             for i in range(len(output_record[s1])):
                 if self.scale_nominal_param_value:
```
Add comments to explain this. Why the if statement? Mention the finite difference formula we are using.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -1011,18 +1104,25 @@ def __create_doe_model(self):
+        def flatten_time(m, j):
+            return self.flatten_measure_timeset[j]
+        #m.tmea_set = Param(m.y_set, initialize=flatten_time, within=Reals)
+        # TODO: expand this to any y set
```
Let's discuss how to incorporate this into redesigning the measurement class.

adowling2 reviewed (commented):
@jialuw96 We should meet and revisit our notes on how to design the measurement class. I recall we worked through all of the likely cases a month ago. We should also discuss how to move as much of the flattening logic as possible to the measurement class.

jialuw96 commented:
![894350B2-AC9F-40AD-B856-B9E2C7F8045D_1_105_c](https://user-images.githubusercontent.com/51490104/142770470-e8a1e68d-5748-48a1-81ea-77533e61c082.jpeg)

jialuw96 commented:
Changes this time:
- the measurement class is created by users
- a check_subset() function is added to the measurement class to check if the subset is correctly defined

Notes for measurement class 
What we need is: 1. measure a big measurement set with all measurements; 2. calculate Jacobian and FIM with a select subset of measurement. 
2 is accomplished in the FIM_result class. This class is created containing the whole Jacobian with all measurements. The user can use this class and call the function calculate_FIM giving the subset of measurements, i.e. FIM_result.calculate_FIM({measurement subset}). When changing subsets, what needs to be done is call this function multiple times with different inputs, but only one FIM_result class is needed.

jialuw96 commented:
![9467D40C-1E93-478B-82CD-F1F99957577F_1_105_c](https://user-images.githubusercontent.com/51490104/144154440-b55603e1-275e-41e7-b8b9-aa367ef9afd3.jpeg)

jialuw96 commented:
@adowling2 Please see the updated code for the FIM_Result updates. It is now splitted to two classes. Thank you! 

- I always ask the DOE class to take all measurements it can take, and give an overall Jacobian matrix (except for simultaneous mode) 
- I used to use FIM class to take the overall Jacobian, and use functions in FIM class to take different measurements, giving back different FIM. In this function, I divided the Jacobian as needed. 
- If I am going to specify an individual FIM class for each measurement subset, I need to think about where I store the overall Jacobian, and where I split it. 

I split FIM_result into two classes: 
Jac_splitter: store the overall Jacobian, can split a jacobian matrix according to different measurement class. 
FIM_result: takes a measure object, prior information; give this measure object to jac_splitter() to get the corresponding Jacobian information; calculate FIM information

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,13 +6,131 @@
+        print('Flatten measurement timeset:', self.flatten_measure_timeset)
+        return self.flatten_measure_name, self.flatten_measure_timeset
+
+    def generate_flatten_name(self, measure_name_and_index):
```
Private function. Use `self.measure_name_and_index`.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,13 +6,131 @@
+            else:
+                flatten_names.append(j)
+
+        return flatten_names
```
Instead of returning, store results in `self.flatten_names`.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,13 +6,131 @@
+        flatten_measure_timeset: a dictionary, keys are flattened measurement name, values are lists of measurement timepoints
+        '''
+        # check if measurement variables need to be flattened
+        self.name_and_index = self.name_and_index_generator(self.measurement_all_info)
```
`self.__name_and_index_generator()`

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,13 +6,131 @@
+        return flatten_names
+
+    def generate_flatten_timeset(self, all_info, flatten_measure_name,name_and_index):
+        '''
```
Same comments as previous function.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,13 +6,131 @@
+        TODO: return True or False . Optional argument: throw error = False
+        subset: measurement name and index involved in jacobian calculation
+        '''
+        flatten_subset = self.generate_flatten_name(subset)
```
Instead keep `subset` to be a Measurement object. Directly access these elements of `subset`.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,13 +6,131 @@
+        flatten_name_index = self.name_and_index_generator(subset)
+        flatten_timeset = self.generate_flatten_timeset(subset, flatten_subset, flatten_name_index)
+        for i in flatten_subset:
+            if i not in self.flatten_measure_name:
```
Add comments. Verify every flattened measurement name in `subset` is contained in `self`.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,13 +6,131 @@
+                raise ValueError('This is not a legal subset of the measurement overall set!')
+            else:
+                for t in flatten_timeset[i]:
+                    if t not in self.flatten_measure_timeset[i]:
```
Add comment here.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -23,9 +141,7 @@ def __init__(self, param_init, design_variable_timepoints, measurement_variable_
-        measurement_variable_timepoints: a dictionary, keys are measurement variable names, value are its measuring time points
-        measurement_variables: the variable name of the model output, for e.g., ['CA', 'CB', 'CC'].
-        measurement_timeset: a list of measurement time points. can be different from control time points
+        measurement_index_and_time: the measurement object
```
Change variable name to `measurements`

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -49,32 +164,15 @@ def __init__(self, param_init, design_variable_timepoints, measurement_variable_
-        print('The flattend measurements include:', self.flatten_measure_name)
+        # create the measurement information object
+        self.measure = measurement_object
+        self.flatten_measure_name, self.flatten_measure_timeset = self.measure.flatten_measurement()
```
Instead the __init__ method in `Measurement` should automatically flatten, store results in `measurement.flat_names` or `measurement.flat_timeset`.

adowling2 replied:
`self.measure.flat_names`.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -1611,46 +1828,361 @@ def create_scenario(self):
 
 class FIM_result:
-    def __init__(self, para_name, measurement_variables, measurement_timeset,  prior_FIM=None, store_FIM=None, scale_constant_value=1, max_condition_number=1.0E12,
+    def __init__(self, para_name, measure_object, jacobian_splitter, prior_FIM=None, store_FIM=None, scale_constant_value=1, max_condition_number=1.0E12,
```
Alternate design: replace `jacobian_splitter` with `jac_3d`.

adowling2 reviewed (commented):
Feedback from the meeting today.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,13 +6,127 @@
+        Check if the subset is correctly defined with right name, index and time.
+        subset: measurement name and index involved in jacobian calculation
+        throw_error: if the given subset is not a subset of the measurement set, throw error message
+        '''
```
Add to docstring.

Arguments:
    subset: description here

Returns:
    valid_subset: Boolean, true if `subset` is contained in `self`

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -394,22 +534,34 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
+                    for j in self.flatten_measure_name:
+                        # split flattened name if needed
+                        if '_index_' in j:
+                            measure_name = j.split('_index_')[0]
```
self.measure.__index_string

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,13 +6,127 @@
+        # a list of measurement names
+        self.measurement_name = list(measurement_index_time.keys())
+        # begin flatten
+        self.__name_and_index_generator(self.measurement_all_info)
```
self.__index_string = '_index_'

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -6,13 +6,127 @@
+        for j in list(measure_name_and_index.keys()):
+            if measure_name_and_index[j] is not None: # if it has extra index
+                for ind in measure_name_and_index[j]:
+                    flatten_name = j + '_index_' + str(ind)
```
self.__index_string

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -394,22 +534,34 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
+                        dataframe.to_csv(mod_name)
+
                     # loop over measurement item and time to store model measurements
                     output_combine = []
```
Make this a function. extract_measurement_values

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -642,29 +838,55 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
-                        try:
+            for mname in self.flatten_measure_name:
+                if '_index_' in mname:
+                    measure_name = mname.split('_index_')[0]
```
Replace _index_

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -632,8 +825,11 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
             dsdp_array = dsdp_re.toarray().T
             # here for construction. Remove after finishing.
             dd = pd.DataFrame(dsdp_array)
```
Toggle on only if verbose. This appears to be left over from debugging.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -642,29 +838,55 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
+                    measure_index = mname.split('_index_')[1]
+                    for tim in measurement_accurate_time[mname]:
+                        # get the measurement name in the model
+                        measurement_name = measure_name+'[0,'+measure_index+','+str(tim)+']'
```
Why not put this in the measurement class?

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -642,29 +838,55 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
+                else:
+                    for tim in self.flatten_measure_timeset[mname]:
+                        # get the measurement name in the model
+                        measure_name = mname+'[0,'+str(tim)+']'
```
Move this into the measurement class.

adowling2 replied:
Ask the measurement class to return the name for the Pyomo models. This will remove duplicate code here. It will also make it easier to test.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -706,41 +928,75 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
         return jac
 
+    def __extract_jac(self, m, y_set=None, t_all_set=None):
+        '''
```
Update. yset = self..., t_all_set = self...

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -706,41 +928,75 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
+        #                JAC[n1, n2, t] = value(m.jac[name1, name2, tim])
+                        
+        jac = {}
+        for p in self.param_name: 
```
Add comments

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -1115,15 +1387,24 @@ def jac_numerical(m,j,p,t):
+                if type(self.measure.name_and_index[measure_name][0]) is str:
+                    measure_index = '"' + measure_index + '"'
+                if t in self.flatten_measure_timeset[j]:
+                    up_C = eval('m.' + measure_name + '[' + str(scenario_all['jac-index'][p][0]) + ',' + measure_index + ',' + str(t) + ']')
```
Can the measurement class give us the Pyomo string name? Here we would just provide `str(scenario_all['jac-index'][p][0])` as an input to a function in the measurement class.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -1613,44 +1897,50 @@ def create_scenario(self):
-        measurement_timeset: measurement variable control timeset
-        measurement_extra_index: measurement indexes other than time
+        measure_object: measurement information object
+        jacobian_info: the jacobian for this measurement object
```
Proposal: only keep one Jacobian (which corresponds with measure_object).

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -1728,80 +1993,78 @@ def calculate_FIM(self, jaco_information, dv_values, jaco_involved_name=None, ja
+        small_jac = self.__split_jacobian(measurement_subset)
+
+        # create a new subject
+        FIM_subclass = FIM_result(self.para_name, measurement_subset, jacobian_info=small_jac, all_jacobian_info=self.all_jacobian_info, prior_FIM=self.prior_FIM, store_FIM=self.store_FIM, scale_constant_value=self.scale_constant_value, max_condition_number=self.max_condition_number)
```
Only pass the small Jacobian to the new FIM_result object.

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -1728,80 +1993,78 @@ def calculate_FIM(self, jaco_information, dv_values, jaco_involved_name=None, ja
-        self.__solution_info(m, dv_set)
-        self.__get_solver_info()
+        # convert the form of jacobian for split
+        jaco_3D = self.__jac_reform_3D(self.all_jacobian_info)
```
Update to self.jacobian

adowling2 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -1728,80 +1993,78 @@ def calculate_FIM(self, jaco_information, dv_values, jaco_involved_name=None, ja
+        jaco_3D = self.__jac_reform_3D(self.all_jacobian_info)
+
+        involved_flatten_index = measurement_subset.flatten_measure_name
+        print('involved flatten name:', involved_flatten_index)
```
Use verbose toggle

adowling2 reviewed (commented):
See comments.

jialuw96 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -92,17 +102,79 @@ def __generate_flatten_timeset(self, all_info, flatten_measure_name,name_and_ind
 
-    def check_subset(self,subset, throw_error=True):
+    def __model_measure_name(self):
+        measurement_names = []
```
Doc string

jialuw96 replied:
Add an example to the doc string

jialuw96 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -92,17 +102,79 @@ def __generate_flatten_timeset(self, all_info, flatten_measure_name,name_and_ind
+    def __model_measure_name(self):
+        measurement_names = []
+        for mname in self.flatten_measure_name:
+            if self.ind_string in mname:
```
Need to add comments

jialuw96 replied:
Split into measurement name and extra index

jialuw96 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -92,17 +102,88 @@ def __generate_flatten_timeset(self, all_info, flatten_measure_name,name_and_ind
+                    measurement_name = measure_name + '[0,' + measure_index + ',' + str(tim) + ']'
+                    measurement_names.append(measurement_name)
+            else:
+                for tim in self.flatten_measure_timeset[mname]:
```
if no extra index

jialuw96 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -92,17 +102,88 @@ def __generate_flatten_timeset(self, all_info, flatten_measure_name,name_and_ind
+            if self.ind_string in j:
+                measure_name = j.split(self.ind_string)[0]
+                measure_index = j.split(self.ind_string)[1]
+                if type(self.name_and_index[measure_name][0]) is str:
```
Add comments

jialuw96 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -92,17 +102,88 @@ def __generate_flatten_timeset(self, all_info, flatten_measure_name,name_and_ind
+        if mode=='simultaneous_finite':
+            # check extra index
+            if self.ind_string in j:
+                measure_name = j.split(self.ind_string)[0]
```
Store this in self? This code is repeated.

jialuw96 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -92,17 +102,88 @@ def __generate_flatten_timeset(self, all_info, flatten_measure_name,name_and_ind
+                measure_name = j.split(self.ind_string)[0]
+                measure_index = j.split(self.ind_string)[1]
+                if type(self.name_and_index[measure_name][0]) is str:
+                    measure_index = '"' + measure_index + '"'
```
Remove repetition

jialuw96 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -92,17 +102,88 @@ def __generate_flatten_timeset(self, all_info, flatten_measure_name,name_and_ind
+                measure_index = j.split(self.ind_string)[1]
+                if type(self.name_and_index[measure_name][0]) is str:
+                    measure_index = '"' + measure_index + '"'
+                if t in self.flatten_measure_timeset[j]:
```
model_name = 'm.'

jialuw96 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -92,17 +102,88 @@ def __generate_flatten_timeset(self, all_info, flatten_measure_name,name_and_ind
+                if type(self.name_and_index[measure_name][0]) is str:
+                    measure_index = '"' + measure_index + '"'
+                if t in self.flatten_measure_timeset[j]:
+                    string_name = 'mod.' + measure_name + '[0,' + str((measure_index)) + ',' + str(t) + ']'
```
model_name = 'mod.'

jialuw96 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -676,9 +699,9 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
-                            measure_name = j.split('_index_')[0]
-                            measure_index = j.split('_index_')[1]
+                        if self.measure.ind_string in j:
+                            measure_name = j.split(self.measure.ind_string)[0]
```
Instead store these in self.measure

jialuw96 commented on pyomo/contrib/doe/fim_doe.py:
```diff
@@ -837,56 +861,28 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
+                if self.discretize_model is not None:
+                    # for DAE model, some variables are fixed
+                    try:
+                        kaug_no = col.index(measurement_name)
```
Use string name to locate the column, extra information

jialuw96 reviewed (commented):
Meeting with @adowling2

jialuw96 commented:
@adowling2 Ready for review: 
- fim_doe.py (Update variance)
- Regression_test.py (the file ends with .py, not with .ipynb)
- test.py (testing measurement and parameter class, ending with .py, not with .ipynb)
Thank you! Jialu

adowling2 commented on pyomo/contrib/doe/Regression_test.ipynb:
```diff
@@ -0,0 +1,201 @@
+{
```
You'll want to remove this file before merging your PR.

adowling2 commented on pyomo/contrib/doe/Regression_test.py:
```diff
@@ -0,0 +1,119 @@
+
```
Add the copyright notice here.

adowling2 commented on pyomo/contrib/doe/Regression_test.py:
```diff
@@ -0,0 +1,119 @@
+
+
+class doe_object_Tester(unittest.TestCase):
+    def setUP(self):
```
Add a docstring here to better explain this class.

adowling2 commented on pyomo/contrib/doe/design_of_exp_brute_v2.ipynb:
```diff
@@ -284,7 +284,7 @@
     "collo = False\n",
     "exp_time = time_points(NFEt, time_scale)\n",
     "\n",
-    "scena_all = {'scena-name': [0], 'fitted_transport_coefficient':212, 'ua': np.log(5.0E6)}\n",
```
Remove this file.

adowling2 commented on pyomo/contrib/doe/draw_figures.ipynb:
```diff
@@ -7,31 +7,3941 @@
    "metadata": {},
```
Remove this file? Instead move to CCSI repo?

adowling2 commented on pyomo/contrib/doe/reactor_kinetics.py:
```diff
@@ -0,0 +1,217 @@
+from pyomo.environ import *
```
Add copyright notice to this file.

adowling2 commented on pyomo/contrib/doe/reactor_kinetics.py:
```diff
@@ -0,0 +1,217 @@
+import numpy as np
+
+def discretizer(m, NFE=32):
+    discretizer = TransformationFactory('dae.collocation')
```
Provide a doc string for all functions in this file.

adowling2 commented on pyomo/contrib/doe/reactor_kinetics.py:
```diff
@@ -0,0 +1,217 @@
+def discretizer(m, NFE=32):
+    discretizer = TransformationFactory('dae.collocation')
+    discretizer.apply_to(m, nfe=NFE, ncp=3, wrt=m.t)
+    #m = discretizer.reduce_collocation_points(m, var=m.T, ncp=1, contset=m.t)
```
Should this be toggled on?

adowling2 commented on pyomo/contrib/doe/reactor_kinetics.py:
```diff
@@ -0,0 +1,217 @@
+    m.k1_pert_rule = Constraint(m.scena, m.t, rule=cal_kp1)
+    m.k2_pert_rule = Constraint(m.scena, m.t, rule=cal_kp2)
+    m.dCdt_rule = Constraint(m.scena,m.y_set, m.t, rule=dCdt_control)
+    #m.dCdt_rule.deactivate()
```
Remove this code?

adowling2 commented on pyomo/contrib/doe/test.ipynb:
```diff
@@ -0,0 +1,233 @@
+{
```
Remove this file.

adowling2 commented on pyomo/contrib/doe/test.py:
```diff
@@ -0,0 +1,56 @@
+from pyomo.environ import *
```
Add copyright statement to this file.

adowling2 commented on pyomo/contrib/doe/test.py:
```diff
@@ -0,0 +1,56 @@
+from fim_doe import *
+
+class TestMeasurement(unittest.TestCase):
+    
```
Add doc string to explain the purpose of this class in a few sentences.

adowling2 commented on pyomo/contrib/doe/test.py:
```diff
@@ -0,0 +1,56 @@
+        
+
+class TestParameter(unittest.TestCase):
+    
```
Add doc string.

adowling2 reviewed (changes requested):
@jialuw96 I finished my review. You want to remove all of the unneeded files. Specifically, I think you only want a handful of files:
- `fim_doe.py`
- One .py file containing all of the necessary reactor example models
- `test_fim_doe.py`
- `test_reactor_example.py`
- `fim_doe_tutorial.ipynb` - A single notebook that walks the user through how to some the reactions example and reproduce key plots from the paper.

jialuw96 commented:
@adowling2 Files are organized and cleaned up here. This branch is ready to be merged.

adowling2 commented:
Looks good. Let's merge.
